syntax = "proto3";

package store;

service KeyValueStore {
  rpc put(PutRequest) returns (PutResponse);
  rpc get(GetRequest) returns (GetResponse);
  rpc slowDown(SlowDownRequest) returns (SlowDownResponse);
  rpc restore(RestoreRequest) returns (RestoreResponse);

  rpc votePut(Empty) returns (Vote);
  rpc commitPut(PutRequest) returns (PutResponse);
  rpc voteGet(Empty) returns (Vote);

  rpc restoreData(Empty) returns (RestoreDataResponse);

  rpc registerNode(NodeParams) returns (RegisterNodeResponse);

  rpc discoverMe(NodeParams) returns (DiscoverMeResponse);

  rpc ping(Ping) returns (Pong);
}

message PutRequest {
  string key = 1;
  string value = 2;
}

message PutResponse {
  bool success = 1;
}

message GetRequest {
  string key = 1;
}

message GetResponse {
  string value = 1;
  bool found = 2;
}

message SlowDownRequest {
  int32 seconds = 1;
}

message SlowDownResponse {
  bool success = 1;
}

message Vote {
  bool success = 1;
}

message RestoreRequest {}
message Empty {}

message RestoreResponse {
  bool success = 1;
}

message RestoreDataResponse {
  map<string, string> data = 1;
}

message NodeParams { 
  int32 id = 1;
  string ip = 2;
  int32 port = 3;
  int32 weight = 4;
}

message RegisterNodeResponse { 
  bool success = 1;
}


message Ping { 
  bool success = 1;
}

message Pong { 
  bool success = 1;
}

message DiscoverMeResponse {
  bool success = 1;
  int32 id = 2;
  int32 weight = 3;
  map<string, int32>  quorum = 4;
  map<int32, NodeParams> node_dict = 5;
}

